File: src/main/scala/com/learning/datastructures/linkedlist/SinglyLinkedList.scala
package com.learning.datastructures.linkedlist

import com.learning.datastructures.linkedlist.SinglyLinkedList.Node

class SinglyLinkedList[T] {
  private var currentSize: Int = 0
  private var head: Node[T] = _

  // O(1) for insertions in the beginning of the list
  def add(e: T): SinglyLinkedList[T] = {
    val newHead = Node(e, head)
    head = newHead
    currentSize = currentSize + 1
    this
  }

  // O(1)
  def removeFirst(): T = {
    if (head == null) {
      throw new IndexOutOfBoundsException("Can not remove from an empty list")
    }

    val removed = head.value
    head = head.next
    currentSize = currentSize - 1

    removed
  }

  def remove(i: Int): T = {
    if (i > size || head == null) {
      throw new IndexOutOfBoundsException(s"Can not remove $i element from list with size $size")
    } else if (i == 0) {
      removeFirst()
    } else {
      var p: Node[T] = head
      for (_ <- 0 until i - 1) {
        p = p.next
      }
      val toBeRemoved = p.next.value

      p.next = p.next.next
      currentSize = currentSize - 1

      toBeRemoved
    }
  }

  // O(n)
  def get(i: Int): T = {
    var e: Node[T] = head
    for (_ <- 0 until i) {
      e = e.next
    }
    e.value
  }

  // O(1)
  def size: Int = {
    currentSize
  }

  // O(n)
  override def equals(other: Any): Boolean = {
    other match {
      case otherList: SinglyLinkedList[T] if this.currentSize == otherList.currentSize =>
        var isEqual = true
        for (i <- 0 until currentSize) {
          isEqual = isEqual && get(i) == otherList.get(i)
        }
        isEqual
      case _ => false
    }
  }

  // O(n)
  override def hashCode: Int = {
    var hashCode = 0
    for (i <- 0 until currentSize) {
      hashCode = hashCode * 31 + get(i).hashCode()
    }
    hashCode
  }
}

object SinglyLinkedList {
  def apply[T](xs: T*): SinglyLinkedList[T] = {
    val list = new SinglyLinkedList[T]
    xs.reverse.foreach(list.add)
    list
  }

  def from[T](xs: Seq[T]): SinglyLinkedList[T] = {
    apply(xs: _*)
  }

  case class Node[T](value: T, var next: Node[T])

}

File: src/main/java/com/learning/datastructures/linkedlist/SinglyLinkedList.java
package com.learning.datastructures.linkedlist;

public class SinglyLinkedList<T> implements LinkedList<T> {
    private Node<T> head;
    private Node<T> tail;
    private int size;

    /**
     * Time complexity - O(n)
     * Space complexity - O(n)
     */
    @SafeVarargs
    public static <T> SinglyLinkedList<T> of(T... elements) {
        SinglyLinkedList<T> list = new SinglyLinkedList<>();

        for (int i = elements.length - 1; i >= 0; i--) {
            list.addFirst(elements[i]);
        }

        return list;
    }

    public SinglyLinkedList() {
        size = 0;
    }

    /**
     * Time complexity - O(1)
     * Space complexity - O(1)
     */
    @Override
    public void addFirst(T e) {
        Node<T> newNode = new Node<>(e);
        if (head == null) {
            head = newNode;
            tail = head;
        } else {
            newNode.next = head;
            head = newNode;
        }
        size++;
    }

    /**
     * Time complexity - O(1)
     * Space complexity - O(1)
     */
    @Override
    public void addLast(T e) {
        Node<T> newNode = new Node<>(e);
        if (tail == null) {
            head = newNode;
            tail = head;
        } else {
            tail.next = newNode;
            tail = tail.next;
        }
        size++;
    }

    /**
     * Time complexity - O(1)
     * Space complexity - O(1)
     */
    @Override
    public T getFirst() {
        return head == null ? null : head.value;
    }

    /**
     * Time complexity - O(1)
     * Space complexity - O(1)
     */
    @Override
    public T getLast() {
        return tail == null ? null : tail.value;
    }

    /**
     * Time complexity - O(1)
     * Space complexity - O(1)
     */
    @Override
    public T removeFirst() {
        if (head == null) return null;

        T value = head.value;
        head = head.next;
        size--;
        if (size == 0) tail = null;

        return value;
    }

    /**
     * Time complexity - O(n)
     * Space complexity - O(1)
     */
    @Override
    public T removeLast() {
        if (tail == null) return null;

        T value = tail.value;
        size--;
        if (size == 0) {
            head = null;
            tail = null;
        } else {
            Node<T> current = head;
            while (current.next != tail) {
                current = current.next;
            }
            current.next = null;
            tail = current;
        }

        return value;
    }

    /**
     * Time complexity - O(1)
     * Space complexity - O(1)
     */
    @Override
    public int size() {
        return size;
    }

    /**
     * Time complexity - O(1)
     * Space complexity - O(1)
     */
    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    private static class Node<T> {
        T value;
        Node<T> next;

        Node(T value) {
            this.value = value;
        }
    }
}
